### PR TITLE
EZP-28055: As a v2 Editor I want to search for content items by entering a search keyword

### PR DESCRIPTION
**Pull request by mikadamczyk**
> https://jira.ez.no/browse/EZP-28055

This PR implements a user search for content item in the content repository. He can initiate a search by entering a search keyword and the application will retrieve a list of all the content items that have the same keyword in any of their searchable fields. This search feature use a PHP API along with a dedicated Symfony controller. The results is sorted by modified. As paginator `pagerfanta/pagerfanta` was used.

### CODE REVIEW

File: src/lib/Form/Type/Search/SimpleSearchType.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            'csrf_protection' => false,
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getBlockPrefix()
+    {
+        return null;
```
Comment (webhdx) [inline src/lib/Form/Type/Search/SimpleSearchType.php]: Why do you set it to `null`? Block prefix is crucial for form theming. We do not have any strong convention but I started using underscored class name with type replaced with form word: `simple_search_form`.

File: src/lib/Form/Type/Search/SimpleSearchType.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class SimpleSearchType extends AbstractType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->setMethod('GET')
```
Comment (webhdx) [inline src/lib/Form/Type/Search/SimpleSearchType.php]: I'm not sure if we should force method in form builder. This kills reusability. You can always change this on form creation.

File: src/lib/Form/Type/Search/SimpleSearchType.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults([
+            'data_class' => SimpleSearchData::class,
+            'csrf_protection' => false,
```
Comment (webhdx) [inline src/lib/Form/Type/Search/SimpleSearchType.php]: See my comment above.

File: src/bundle/Resources/views/admin/search/search_form.html.twig [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{{ form_start(form) }}
+<div class="input-group">
+    {{ form_widget(form.query, {'attr': {'style': 'width: 200px', 'class': 'form-control'} }) }}
```
Comment (webhdx) [inline src/bundle/Resources/views/admin/search/search_form.html.twig]: We have to move inline `style` to CSS. @dew326 can propose you best solution.

File: src/bundle/Resources/views/admin/search/search.html.twig [hunk 5]
```diff
@@ -0,0 +1,32 @@
+{% extends "EzPlatformAdminUiBundle::layout.html.twig" %}
+
+{% trans_default_domain 'search' %}
+
+{% block pageTitle %}{% endblock %}
+
+{% block content %}
+    <div id="react-udw"></div>
```
Comment (webhdx) [inline src/bundle/Resources/views/admin/search/search.html.twig]: Should UDW be a part of search page?

File: src/bundle/Controller/SearchController.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                new ContentSearchAdapter(
+                    $query,
+                    $this->searchService
+                )
+            );
+
+            $pagerfanta->setMaxPerPage($limit);
+            $pagerfanta->setCurrentPage($page);
+
+            $urlGenerator = $this->get('router');
```
Comment (webhdx) [inline src/bundle/Controller/SearchController.php]: Please inject in the constructor. We tend to not use container directly in the controllers.

File: src/bundle/Controller/SearchController.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                return $urlGenerator->generate('ezplatform.search', [
+                    'page' => $page,
+                    'limit' => $data->getLimit(),
+                    'query' => $data->getQuery(),
+                ]);
+            };
+
+            $pagination = (new TwitterBootstrap4View())->render($pagerfanta, $routeGenerator);
+
+            return $this->render('@EzPlatformAdminUi/admin/search/list.html.twig', [
```
Comment (webhdx) [inline src/bundle/Controller/SearchController.php]: Can't we share the same template file for search? By looking at the twig files I can see there is a lot of code duplication going on.

File: src/bundle/Resources/views/admin/search/search.html.twig (hunk 5 above)
Comment (dew326) [reply src/bundle/Resources/views/admin/search/search.html.twig]: This is something we should change I think. We load UDW module in every page (in layout.html.twig) but the container is added in some subpage. I think we should move this container to the `layout.html.twig` and remove it from every subpage.

File: src/bundle/Resources/views/admin/search/search.html.twig [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            </section>
+        </div>
+    </div>
+{% endblock %}
+
+{% block title %}{{ 'search.perform'|trans }}{% endblock %}
+
+{% block javascripts %}
+    {% javascripts
+    'bundles/ezplatformadminui/js/scripts/udw/browse.js' %}
```
Comment (dew326) [inline src/bundle/Resources/views/admin/search/search.html.twig]: This JS is always needed with the sidebar, if we have sidebar in `layout.html.twig` this JS should be loaded in `layout.html.twig`

File: src/bundle/Resources/views/layout.html.twig [hunk 9]
```diff
@@ -48,7 +48,32 @@
...
+                            <svg class="ez-icon ez-icon-search">
+                                <use xlink:href="{{ asset('bundles/ezplatformadminui/img/ez-icons.svg') }}#search"></use>
+                            </svg>
+                            {{ 'navigation.admin.search'|trans|desc('Search') }}
+                        </a>
+                        <button class="btn btn-dark btn-block btn--udw-browse" data-starting-location-id="2">
+                            <svg class="ez-icon ez-icon-browse">
+                                <use xlink:href="{{ asset('bundles/ezplatformadminui/img/ez-icons.svg') }}#browse"></use>
+                            </svg>
+                            Browse
```
Comment (dew326) [inline src/bundle/Resources/views/layout.html.twig]: Translation?

File: src/bundle/Resources/views/admin/search/search_form.html.twig [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+
+    <span class="input-group-btn">
+        <input type="submit" class="btn btn-primary" value="Search"/>
```
Comment (dew326) [inline src/bundle/Resources/views/admin/search/search_form.html.twig]: I think it should be `<button>`, and I think the icon is missing in this input.

File: src/bundle/Controller/SearchController.php (hunk 7 above)
Comment (mikadamczyk) [reply src/bundle/Controller/SearchController.php]: Yes we can, but the grid template will be added, so I prefer separate templates to avoid mixing. What do you think?

File: src/lib/Form/Type/Search/SimpleSearchType.php (hunk 1 above)
Comment (mikadamczyk) [reply src/lib/Form/Type/Search/SimpleSearchType.php]: I used it because I want simple parameters in the URL. I don't think so the block prefix is crucial for form theming, rather id or class added in twig template. I think form builder is not a good place for styling. What do you think?

File: src/bundle/Resources/views/admin/search/search_form.html.twig (hunk 4 above)
Comment (mikadamczyk) [reply src/bundle/Resources/views/admin/search/search_form.html.twig]: Inline style is not needed anymore

File: src/lib/Form/Type/Search/SimpleSearchType.php (hunk 1 above)
Comment (webhdx) [reply src/lib/Form/Type/Search/SimpleSearchType.php]: Hmm, I'm not sure if you understand my point. All forms should have unique block name, this allows you to do overrides in twig:

```twig
{% block simple_search_form %}
    {% set attr = attr|merge({'class': 'custom-class'}) %}
    {{ block('form_widget') }}
{% endblock %}
```

Which is what we do in order to separate styling information (like `class` HTML attribute) from PHP code.

File: src/bundle/Resources/views/layout.html.twig (hunk 9 above)
Comment (webhdx) [reply src/bundle/Resources/views/layout.html.twig]: It's going to be fixed in #26, @mikadamczyk can leave this as-is since he only moved the code.

File: src/bundle/Resources/views/admin/search/search.html.twig (hunk 5 above)
Comment (lserwatka) [reply src/bundle/Resources/views/admin/search/search.html.twig]: This is good point @dew326 . We should make a follow-up task on this UDW load.

File: src/bundle/Controller/SearchController.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            /** @var SearchData $data */
+            $data = $form->getData();
+
+            $query = new Query();
+            $query->filter = new Criterion\LogicalAnd(
+                [
+                    new Criterion\Visibility(Criterion\Visibility::VISIBLE),
+                    new Criterion\FullText($data->getQuery()),
```
Comment (alongosz) [inline src/bundle/Controller/SearchController.php]: Please validate if query is string and if it's not empty (to avoid making unnecessary calls to external SE handler).

File: src/bundle/Controller/SearchController.php [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+                ]
+            );
+            $query->sortClauses[] = new SortClause\DateModified(Query::SORT_ASC);
+
```
Comment (alongosz) [inline src/bundle/Controller/SearchController.php]: Please actually use `$limit` for query (`$query->limit`). Otherwise it will return default number of records (AFAICS now it is 25).
Also there's a need for `$query->offset`.

I don't know if and how this works with Pagerfanta, but offset and limits are needed to avoid killing external SE handler (e.g. can be Solr on a separate server)

File: composer.json [hunk 13]
```diff
@@ -24,9 +24,6 @@
     "require-dev": {
         "phpunit/phpunit": "^6.4"
     },
-    "require-dev": {
-        "phpunit/phpunit": "^6.4"
-    },
```
Comment (alongosz) [inline composer.json]: Changes out of scope should be in separate commits (yeah, we usually squash, but sometimes we don't).

File: src/lib/Form/Data/Search/SearchData.php [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+class SearchData
+{
+    /** @var int */
+    private $limit;
+
+    /** @var int */
+    private $page;
+
+    /** @var string */
+    private $query;
```
Comment (alongosz) [inline src/lib/Form/Data/Search/SearchData.php]: I never quite understood the need of encapsulation of data structure properties just to write getters and setters getting and setting them directly. This make sense to me only if structure is supposed to be immutable.

But not requesting changes, just asking why this is better. The flame war about this started around 2003 and never ended :)

Comment (webhdx) [reply src/lib/Form/Data/Search/SearchData.php]: Accessing properties directly looks kinda lame. `$something->getProperty()` looks sexier. 😂

Comment (mikadamczyk) [reply src/lib/Form/Data/Search/SearchData.php]: I think the only reason is that other classes are the same :(

**Comment by lserwatka** [comment]
@webhdx @alongosz are you OK now?

**Comment by alongosz** [comment]
Nitpicks: 
- still would prefer change of composer.json which is out of scope to be in a separate commit, but if it's gonna be squashed, then there's no point :)
- also commit message (after expanding) looks a little bit weird, but this can be handled on merge if squashed

File: composer.json [hunk 15]
```diff
@@ -22,9 +22,6 @@
         "ezsystems/ez-support-tools": "^0.1@dev",
         "knplabs/knp-menu-bundle": "^2.1"   
     },
-    "require-dev": {
```
Comment (webhdx) [inline composer.json]: @mikadamczyk Why this PR removes phpunit? Seems unrelated/bad?

File: src/bundle/Controller/SearchController.php [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+use eZ\Publish\Core\Pagination\Pagerfanta\ContentSearchAdapter;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use EzSystems\EzPlatformAdminUi\Form\Factory\FormFactory;
+use EzSystems\EzPlatformAdminUi\Form\SubmitHandler;
+
+class SearchController extends Controller
+{
+    /** @var SearchService */
+    private $searchService;
```
Comment (webhdx) [inline src/bundle/Controller/SearchController.php]: I would define all properties either `private` or `protected`. `private` seems to be a more popular choice with recent PRs in order to disallow relying someone's code on our controllers.

File: src/lib/Form/Type/Search/SearchType.php [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+use Symfony\Component\Form\Extension\Core\Type\SearchType as CoreSearchType;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class SearchType extends AbstractType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
Comment (webhdx) [inline src/lib/Form/Type/Search/SearchType.php]: Please override `configureOptions()` with default `data_class` set to your `SearchData::class`

File: composer.json (hunk 15 above)
Comment (alongosz) [reply composer.json]: Weird, actually when I reviewed this first it seemed like cleanup of redundant block, but maybe I misread something

File: src/bundle/Controller/SearchController.php [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+                $query = new Query();
+                $query->filter = new Criterion\LogicalAnd(
+                    [
+                        new Criterion\Visibility(Criterion\Visibility::VISIBLE),
+                        new Criterion\FullText($queryString),
+                    ]
+                );
+
+                $query->limit = $limit;
+                $query->offset = ($page - 1) * $limit;
```
Comment (adamwojs) [inline src/bundle/Controller/SearchController.php]: You don't need to set `limit` and `offset` manually, pagerfanta should take care of this.

Comment (alongosz) [reply src/bundle/Controller/SearchController.php]: Ok, so apparently this already works with `ContentSearchAdapter` from kernel (which I didn't know or even expect to be implemented there, sorry). Not 100% convinced this is clean, as being PAPI developer I got confused, but seems like a way to go.

**Comment by barbaragr** [comment]
Performing Search on the second page of search results.
Preconditions: perform search action to receive more than 10 results. I've got 12 elements on a list and 2 pages.

Steps:
1. Go to the second page.
2. Type a word in a search input. 
3. Perform search action - "Page 2 does not exist. The currentPage must be inferior to 1" alert appears and there are no search results. It should automatically go to the first page and show results.

**Comment by barbaragr** [comment]
Searching with phrase containing '%' results with "The expression 'OR' expected at least 1 argument but none provided". Nothing is shown in search result.

**Comment by adamwojs** [comment]
> Searching with phrase containing '%' results with "The expression 'OR' expected at least 1 argument but none provided". Nothing is shown in search result.

@barbaragr This is already known issue. See https://github.com/ezsystems/ezpublish-kernel/pull/2029 for more details.

File: src/bundle/Controller/SearchController.php (hunk 18 above)
Comment (mikadamczyk) [reply src/bundle/Controller/SearchController.php]: Ok, so I will remove `limit` and `offset` for query

Comment (alongosz) [reply src/bundle/Controller/SearchController.php]: yep, sorry for the mess.

**Comment by mikadamczyk** [comment]
I fixed PR, can you take a look at it? @barbaragr @adamwojs @webhdx

**Comment by Nattfarinn** [comment]
@webhdx Please confirm if you're fine with changes to your code review.

File: src/bundle/Resources/views/admin/search/search_form.html.twig [hunk 19]
```diff
@@ -0,0 +1,14 @@
+{{ form_start(form) }}
+<div class="input-group">
+    {{ form_widget(form.query, {'attr': {'class': 'form-control'} }) }}
+    {{ form_widget(form.page, {'attr': {'value': 1 }}) }}
```
Comment (webhdx) [inline src/bundle/Resources/views/admin/search/search_form.html.twig]: We shouldn't overwrite the value in the view. Value can be passed in the controller by changing corresponding data object.
